File: src/lib/api/market.ts
import { invoke } from "@tauri-apps/api/core";
import type { MarketPriceUnit, ProviderMarketModelPrice } from "@/types";

export interface SellerPricingSuggestion {
  pricePer1kTokens: number;
  source: string;
  modelPrice?: ProviderMarketModelPrice;
}

export interface StartSellingTokensInput {
  providerId: string;
  modelName: string;
  // Frontend-facing field, aligned with ProviderSellerConfig semantics.
  pricePer1kTokens: number;
  endpoint: string;
  modelPrices?: ProviderMarketModelPrice[];
  priceUnit?: MarketPriceUnit;
  priceVersion?: number;
  amountFen?: number;
  payTo?: string;
  indicator?: string;
  accessToken?: string;
}

export interface CloudflaredCheckResult {
  installed: boolean;
  version?: string | null;
  path?: string | null;
  installCommand: string;
}

export interface SellerRuntimeStatus {
  providerId: string;
  tunnelRunning: boolean;
  hasActiveToken: boolean;
  status: "idle" | "running";
}

export interface MarketPaymentListing {
  provider: string;
  mode: string;
  amountFen: number;
  currency: string;
  skillSlug: string;
  indicator: string;
  payTo: string;
}

export interface MarketListing {
  provider_id: string;
  model_name: string;
  price_per_1k_tokens: number;
  endpoint: string;
  seller_pubkey: string;
  timestamp: number;
  status?: "available" | "reserved" | "busy" | "offline";
  capacity?: number;
  payment?: MarketPaymentListing | null;
  resourceUrl?: string;
  amountFen?: number;
  accessToken?: string | null;
}

export const marketApi = {
  async checkCloudflared(): Promise<CloudflaredCheckResult> {
    return await invoke("check_cloudflared");
  },

  async startCloudflareTunnel(port: number): Promise<string> {
    return await invoke("start_cloudflare_tunnel", { port });
  },

  async findAiSellers(): Promise<MarketListing[]> {
    return await invoke("find_ai_sellers");
  },

  async startSellingTokens(input: StartSellingTokensInput): Promise<string> {
    return await invoke("start_selling_tokens", {
      input: {
        providerId: input.providerId,
        modelName: input.modelName,
        // Backend command contract expects `price` as the legacy compatibility field.
        price: input.pricePer1kTokens,
        endpoint: input.endpoint,
        modelPrices: input.modelPrices ?? [],
        priceUnit: input.priceUnit ?? "PER_1M_TOKENS",
        priceVersion: input.priceVersion ?? 1,
        amountFen: input.amountFen,
        payTo: input.payTo,
        indicator: input.indicator,
        accessToken: input.accessToken,
      },
    });
  },

  async stopSellingTokens(providerId: string): Promise<boolean> {
    return await invoke("stop_selling_tokens", { providerId });
  },

  async generateSellerAccessToken(providerId: string): Promise<string> {
    return await invoke("generate_seller_access_token", { providerId });
  },

  async getSuggestedSellerPrice(
    providerId: string,
    modelName?: string,
  ): Promise<SellerPricingSuggestion> {
    return await invoke("get_suggested_seller_price", {
      providerId,
      modelName,
    });
  },

  async getSellerRuntimeStatus(
    providerId: string,
  ): Promise<SellerRuntimeStatus> {
    return await invoke("get_seller_runtime_status", { providerId });
  },
};

File: src-tauri/src/commands/market.rs
use crate::services::market::{
    CloudflaredCheckResult, MarketListing, MarketModelPrice, MarketPaymentListing, MarketService,
    SellerPricingSuggestion, SellerRuntimeStatus,
};
use crate::store::AppState;
use serde::Deserialize;
use tauri::State;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StartSellingTokensRequest {
    provider_id: String,
    model_name: String,
    price: u64,
    endpoint: String,
    model_prices: Option<Vec<MarketModelPrice>>,
    price_unit: Option<String>,
    price_version: Option<u32>,
    amount_fen: Option<i64>,
    pay_to: Option<String>,
    indicator: Option<String>,
    access_token: Option<String>,
}

#[tauri::command]
pub async fn start_cloudflare_tunnel(
    state: State<'_, AppState>,
    port: u16,
) -> Result<String, String> {
    log::info!("收到启动 Cloudflare 隧道请求: port={}", port);
    match state.market_service.start_tunnel(port).await {
        Ok(url) => {
            log::info!("Cloudflare 隧道请求完成: port={}, url={}", port, url);
            Ok(url)
        }
        Err(e) => {
            log::error!("Cloudflare 隧道请求失败: port={}, error={}", port, e);
            Err(e.to_string())
        }
    }
}

#[tauri::command]
pub async fn check_cloudflared() -> Result<CloudflaredCheckResult, String> {
    Ok(MarketService::check_cloudflared())
}

#[tauri::command]
pub async fn start_selling_tokens(
    state: State<'_, AppState>,
    input: StartSellingTokensRequest,
) -> Result<String, String> {
    log::info!(
        "收到市场发布请求: provider_id={}, model_name={}, price_per_1k_tokens={}, endpoint={}",
        input.provider_id,
        input.model_name,
        input.price,
        input.endpoint
    );
    let amount_fen = input
        .amount_fen
        .unwrap_or_else(|| input.price.max(1) as i64);
    let indicator = input
        .indicator
        .unwrap_or_else(|| format!("tokens-buddy-{}", input.provider_id));
    let pay_to = input.pay_to.unwrap_or_default();
    let listing = MarketListing {
        provider_id: input.provider_id,
        model_name: input.model_name,
        price_per_1k_tokens: input.price,
        endpoint: input.endpoint.clone(),
        seller_pubkey: "".to_string(), // MarketService will sign it
        timestamp: chrono::Utc::now().timestamp() as u64,
        model_prices: input.model_prices.unwrap_or_default(),
        price_unit: input
            .price_unit
            .unwrap_or_else(|| "PER_1M_TOKENS".to_string()),
        price_version: input.price_version.unwrap_or(1),
        status: crate::services::clawtip::listing::ListingStatus::Available,
        capacity: 1,
        streaming: true,
        payment: Some(MarketPaymentListing {
            provider: "clawtip".to_string(),
            mode: "per_call_prepaid".to_string(),
            amount_fen,
            currency: "CNY_FEN".to_string(),
            skill_slug: "tokens-buddy-llm-console".to_string(),
            indicator,
            pay_to,
        }),
        resource_url: input.endpoint,
        amount_fen,
        access_token: input.access_token,
    };

    match state.market_service.start_selling(listing).await {
        Ok(event_id) => {
            log::info!("市场发布完成: event_id={}", event_id);
            Ok(event_id)
        }
        Err(e) => {
            log::error!("市场发布失败: error={}", e);
            Err(e.to_string())
        }
    }
}

#[tauri::command]
pub async fn find_ai_sellers(state: State<'_, AppState>) -> Result<Vec<MarketListing>, String> {
    state
        .market_service
        .find_sellers()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn stop_selling_tokens(
    state: State<'_, AppState>,
    #[allow(non_snake_case)] providerId: String,
) -> Result<bool, String> {
    log::info!("收到停止售卖请求: provider_id={}", providerId);
    state
        .market_service
        .stop_selling(&providerId)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn generate_seller_access_token(
    #[allow(non_snake_case)] providerId: String,
) -> Result<String, String> {
    log::info!("生成分享访问令牌: provider_id={}", providerId);
    Ok(MarketService::generate_and_register_access_token_for(
        &providerId,
    ))
}

#[tauri::command]
pub async fn get_suggested_seller_price(
    #[allow(non_snake_case)] providerId: String,
    #[allow(non_snake_case)] modelName: Option<String>,
) -> Result<SellerPricingSuggestion, String> {
    Ok(MarketService::suggest_price_for_model(
        &providerId,
        modelName.as_deref(),
    ))
}

#[tauri::command]
pub async fn get_seller_runtime_status(
    state: State<'_, AppState>,
    #[allow(non_snake_case)] providerId: String,
) -> Result<SellerRuntimeStatus, String> {
    Ok(state
        .market_service
        .seller_runtime_status(&providerId)
        .await)
}
